### PR TITLE
Fix GPU restart for pure SoA particles

### PR DESCRIPTION
**Pull request by atmyers**
The proposed changes:
- [x] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Review by ax3l — DISMISSED** [review]
Thank you for debugging and fixing this! :sparkles:

**Comment by ax3l** [comment]
Thank you for debugging and fixing this! ✨

There are a few CI test issues left:
```
0::Assertion `sm_old == sm_new' failed, file "/Users/runner/work/amrex/amrex/Tests/Particles/CheckpointRestartSOA/main.cpp", line 174 !!!
```
and
```
1::Assertion `ptemp.id() > 0' failed, file "/Users/runner/work/amrex/amrex/Src/Particle/AMReX_ParticleIO.H", line 1022 !!!
```

**Comment by WeiqunZhang** [comment]
For Nyx restart, the current development branch works on CPU and GPU. With this branch, it works on CPU, but fails on GPU.

https://ccse.lbl.gov/pub/GpuRegressionTesting/Nyx/2024-03-02/index.html

**Comment by atmyers** [comment]
I just fixed some things, hopefully everything will pass now.

**Comment by WeiqunZhang** [comment]
The Nyx tests pass now. But this IAMR non-restart test still fails. https://ccse.lbl.gov/pub/GpuRegressionTesting/IAMR/2024-03-02-001/Part-2d.html

**Comment by WeiqunZhang** [comment]
The reconnection GPU tests also fail. But the problem might be in the reconnection code, not amrex.

This PR: https://ccse.lbl.gov/pub/GpuRegressionTesting/reconnection/2024-03-02/index.html

development: https://ccse.lbl.gov/pub/GpuRegressionTesting/reconnection/2024-03-01/index.html
The restart test started Friday morning was never able to finish, and I manually killed it today.

@RevathiJambunathan

**Comment by WeiqunZhang** [comment]
The amrex ParticleMesh test fails. https://ccse.lbl.gov/pub/GpuRegressionTesting/AMReX/2024-03-02/ParticleMesh.html

**Comment by WeiqunZhang** [comment]
For reconnection, I also tried an earlier version (that works with the development branch of amrex) with this PR, the errors were similar to those of the latest reconnection code plus this PR. https://ccse.lbl.gov/pub/GpuRegressionTesting/reconnection/2024-03-02-001/index.html

**Comment by RevathiJambunathan** [comment]
Update : I do see the illegal access memory with this current PR, when I ran with 2 GPUs - same as https://ccse.lbl.gov/pub/GpuRegressionTesting/reconnection/2024-03-02-001/Baseline2Dsigma30Perturbation.html


1 . The test where I compiled with the dev branch crashed at the first checkpoint I/O, and this PR fixes that issue 

2. However, when I run the current PR, the test still crashes. Now it write the checkpoint file, but crashes at RedistributeLocal(). Backtrace points to https://github.com/AMReX-Codes/amrex/blob/3525b4a3f27eb64f746dd69b6613f71bb02d6e63/Src/Particle/AMReX_ParticleUtil.H#L588 and https://github.com/AMReX-Codes/amrex/blob/3525b4a3f27eb64f746dd69b6613f71bb02d6e63/Src/Particle/AMReX_ParticleUtil.H#L647

**Comment by AlexanderSinn** [comment]
The issue Reva is seeing was fixed in #3769, the branch of this PR is outdated and hasn't got that fix yet.

**Comment by WeiqunZhang** [comment]
The reconnection code works now.

**Comment by WeiqunZhang** [comment]
Remaining issues:
  - https://ccse.lbl.gov/pub/GpuRegressionTesting/AMReX/2024-03-04-001/ParticleMesh.html
  - https://ccse.lbl.gov/pub/GpuRegressionTesting/IAMR/2024-03-04-001/Part-2d.html

Or is the old behavior actually wrong?

**Comment by WeiqunZhang** [comment]
The issue is in the GPU version of packIOData. If I use managed memory and force it to use the CPU version of packIOData, it works.

**Comment by WeiqunZhang** [comment]
There might be another issue in the GPU version of packIOData besides the one that causes the regression failure. The `IntVector offsets` computed by exclusive sum is never used. If the particle io flags is not all true, that will not be right.

File: Src/Particle/AMReX_WriteBinaryParticleData.H
Comment (WeiqunZhang) [inline Src/Particle/AMReX_WriteBinaryParticleData.H]:
```suggestion
                    const int write_comp_index = j-real_start_offset;
```
?

**Comment by RevathiJambunathan** [comment]
@AlexanderSinn @WeiqunZhang Thanks! 
Yes, updating branch now!

**Comment by WeiqunZhang** [comment]
To summarize the current status, all regression tests I have tried have passed (including the reconnection) except for a few tests in amrex and IAMR after the development branch was merged into this. (Yes, I merged and pushed to this PR.) The amrex and IAMR regressions issues are new issues introduced in this PR. The code change suggestion I made above seems to solve the new regression issues. (No, I did not commit it). There is an existing defect regarding the filtering flags. That should be easy to fix either in this PR or a follow-up.

**Comment by WeiqunZhang** [comment]
To clarify, the reconnection tests work with the current PR and I can approve this PR once the new IAMR and amrex issued are resolved.

**Review by WeiqunZhang — APPROVED** [review]

**Comment by ax3l** [comment]
Just checking for my understanding:
Did WarpX restarts pass now? :)

> There is an existing defect regarding the filtering flags. That should be easy to fix either in this PR or a follow-up.

Is that fixed in the merge or to do?

**Comment by atmyers** [comment]
The WarpX restart issues seem to pass now based on the minimal reproducer. The offsets bug is still to-do.

**Pull request merged**
